node/src/actors/chain_manager/mod.rs

```diff
@@ -829,6 +834,56 @@ impl ChainManager {
         }
     }
 
+    fn update_bit_counter(&mut self, v: u32, epoch: u32) {
+        for n in 0..32i8 {
+            if let Some(mut bit_counter) = self.chain_state.bit_tapi_counter.get_mut(&n, &epoch) {
+                if !self
+                    .chain_state
+                    .wip_activation
+                    .contains_key(&bit_counter.wip)
+                    && is_bit_n_activated(v, n)
+                {
+                    bit_counter.votes += 1;
+
+                    if (epoch + bit_counter.init) % bit_counter.period == 0 {
```

Comment:
Not sure if this calculation is correct. Add a test case:

init = 0, period = 100, should check activation at epoch 100
init = 1, period = 100, should check activation at epoch 101 (or also 100?)

Also, why is `end` ignored here?

Reply:
Its a mistake, it should be epoch - init.
And end is used when you are doing the get_mut, because if you are out of the period, it returns None

Reply:
Does that subtraction check for overflow? When init > epoch this check should always return false.

Reply:
It is not need, because if you are out of the period you can not reach that line